### PR TITLE
CompressionMiddleware: Add Brotli Compression

### PR DESCRIPTION
**Pull request by aidanharris**
This commit adds Brotli compression in addition to the existing GZip
compression. If a client requests either GZip or Brotli compression, Brotli is preferred.

Testing:

Ensure your PHP is configured with the [php-ext-brotli extension](https://github.com/kjdev/php-ext-brotli/). Systems without this extension should continue to work using GZip.

### CODE REVIEW

File: lib/private/AppFramework/Middleware/CompressionMiddleware.php [hunk 1]
```diff
@@ -46,43 +49,58 @@ class CompressionMiddleware extends Middleware {
...
 
-		// Check if we are even asked for gzip
+		// Check if we are even asked for brotli or gzip
 		$header = $this->request->getHeader('Accept-Encoding');
-		if (strpos($header, 'gzip') === false) {
+
+		$allowBrotli = strpos($header, 'br') !== false;
+		$allowGzip = strpos($header, 'gzip') !== false;
+
+		if (($allowGzip === false) && ($allowBrotli === false)) {
```
Comment (kesselb) [inline lib/private/AppFramework/Middleware/CompressionMiddleware.php:70]:
```suggestion
		if ($allowGzip === false && $allowBrotli === false) {
```

File: lib/private/AppFramework/Middleware/CompressionMiddleware.php [hunk 2]
```diff
@@ -46,43 +49,58 @@ class CompressionMiddleware extends Middleware {
...
-			$allowGzip = true;
+			$allowCompression = true;
 		}
 
-		if ($allowGzip) {
-			$this->useGZip = true;
-			$response->addHeader('Content-Encoding', 'gzip');
+		if ($allowCompression) {
+			if ($allowBrotli && function_exists('brotli_compress')) {
+				$this->useGZip = false;
```
Comment (kesselb) [inline lib/private/AppFramework/Middleware/CompressionMiddleware.php:87]:
```suggestion
```

Comment (kesselb) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:87]: false seems the default

**Review by kesselb — COMMENTED** [review]
Thank you very much :+1:

**Review by rullzer — CHANGES_REQUESTED** [review]
I'm not a nhuge fan of such checks. but so be it.
However if I'm not mistaken the RFC specifies to consider the order of the Accept encoding. So we should then properly parse it.

**Comment by kesselb** [comment]
> So we should then properly parse it.

https://symfony.com/doc/current/components/http_foundation.html#accessing-accept-headers-data
https://github.com/symfony/symfony/blob/5.2/src/Symfony/Component/HttpFoundation/AcceptHeader.php
https://github.com/symfony/symfony/blob/5.2/src/Symfony/Component/HttpFoundation/AcceptHeaderItem.php

We use some symfony components but not symfony/http-foundation. However they already solved it properly. So that might be a starting point for us. If the license allows we may include it.

**Comment by kesselb** [comment]
Hi @aidanharris,

We bundle symfony http-foundation since Nextcloud 25. 

Do you have time to look into this feature again?

**Pull request closed without merging**